google-beta/resource_pubsub_subscription.go

```diff
@@ -63,166 +77,375 @@ func resourcePubsubSubscription() *schema.Resource {
-	_, err := call.Do()
+	url, err := replaceVars(d, config, "https://pubsub.googleapis.com/v1/projects/{{project}}/subscriptions/{{name}}")
 	if err != nil {
 		return err
 	}
 
+	var obj map[string]interface{}
+	log.Printf("[DEBUG] Deleting Subscription %q", d.Id())
+	res, err := sendRequestWithTimeout(config, "DELETE", url, obj, d.Timeout(schema.TimeoutDelete))
+	if err != nil {
+		return handleNotFoundError(err, d, "Subscription")
+	}
+
+	log.Printf("[DEBUG] Finished deleting Subscription %q: %#v", d.Id(), res)
 	return nil
 }
 
-func resourcePubsubSubscriptionStateImporter(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+func resourcePubsubSubscriptionImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
 	config := meta.(*Config)
+	if err := parseImportId([]string{"projects/(?P<project>[^/]+)/subscriptions/(?P<name>[^/]+)", "(?P<project>[^/]+)/(?P<name>[^/]+)", "(?P<name>[^/]+)"}, d, config); err != nil {
+		return nil, err
+	}
+
+	// Replace import id for the resource id
+	id, err := replaceVars(d, config, "projects/{{project}}/subscriptions/{{name}}")
+	if err != nil {
+		return nil, fmt.Errorf("Error constructing id: %s", err)
+	}
+	d.SetId(id)
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func flattenPubsubSubscriptionName(v interface{}, d *schema.ResourceData) interface{} {
+	if v == nil {
+		return v
+	}
+	return NameFromSelfLinkStateFunc(v)
+}
+
+func flattenPubsubSubscriptionTopic(v interface{}, d *schema.ResourceData) interface{} {
+	if v == nil {
+		return v
+	}
+	return ConvertSelfLinkToV1(v.(string))
+}
+
+func flattenPubsubSubscriptionLabels(v interface{}, d *schema.ResourceData) interface{} {
+	return v
+}
+
+func flattenPubsubSubscriptionPushConfig(v interface{}, d *schema.ResourceData) interface{} {
+	if v == nil {
+		return nil
+	}
+	original := v.(map[string]interface{})
+	if len(original) == 0 {
+		return nil
+	}
+	transformed := make(map[string]interface{})
+	transformed["push_endpoint"] =
+		flattenPubsubSubscriptionPushConfigPushEndpoint(original["pushEndpoint"], d)
+	transformed["attributes"] =
+		flattenPubsubSubscriptionPushConfigAttributes(original["attributes"], d)
+	return []interface{}{transformed}
+}
+func flattenPubsubSubscriptionPushConfigPushEndpoint(v interface{}, d *schema.ResourceData) interface{} {
+	return v
+}
+
+func flattenPubsubSubscriptionPushConfigAttributes(v interface{}, d *schema.ResourceData) interface{} {
+	return v
+}
+
+func flattenPubsubSubscriptionAckDeadlineSeconds(v interface{}, d *schema.ResourceData) interface{} {
+	// Handles the string fixed64 format
+	if strVal, ok := v.(string); ok {
+		if intVal, err := strconv.ParseInt(strVal, 10, 64); err == nil {
+			return intVal
+		} // let terraform core handle it if we can't convert the string to an int.
+	}
+	return v
+}
 
+func expandPubsubSubscriptionName(v interface{}, d *schema.ResourceData, config *Config) (interface{}, error) {
```

Comment:
Based on https://github.com/terraform-providers/terraform-provider-google-beta/pull/432/files#diff-e50facc7fd1a9dbe302b2930a67034dfR136, we actually only want to return the short name at the end of the api's `name` value; we're deviating from the API a little. You can use https://github.com/GoogleCloudPlatform/magic-modules/blob/master/templates/terraform/custom_expand/resource_from_self_link.go.erb as the expander for this value.

The `name` property shouldn't be present in the API object, but I think it gets ignored at Create time anyways so let's just do that. If there's an error when we provide a malformed value, we can remove it or fix it using an encoder.

Reply:
That was the original plan, but it fails with long resource names that is a behavior we need to preserve. So while the expansion works with `resource_from_self_link` in terms of the field we pass to the API (even if not strictly required), it skips the `d.Set("name", match[2])` below which is needed to ensure the `replaceVars` works with the long name.

Should we just do a `d.Set("name")` if the provided name is in the long form, and skip adding a `name` attribute to the API object in all cases? Would returning `nil` here work for that?

Reply:
Ah, you're correct - I'd missed some of the subtleties on this one. Can you comment above the `d.Set` values why it's important we preserve this behaviour for now?

Reply:
Tested that the `nil` will just pass a nil value to the attribute, but not skip the field.